// @astrojs/cloudflare/dist/index.js
import { createReadStream, existsSync } from "node:fs";
import { appendFile, readFile, rename, stat, unlink, writeFile } from "node:fs/promises";
import { relative } from "node:path";
import { fileURLToPath } from "node:url";
import { normalizePath } from "vite";
import { createInterface } from "node:readline/promises";
import {
  removeLeadingForwardSlash,
  removeTrailingForwardSlash
} from "@astrojs/internal-helpers/path";
import { createRedirectsFromAstroRoutes, printAsRedirects } from "@astrojs/underscore-redirects";
import { cloudflare as cfVitePlugin } from "@cloudflare/vite-plugin";
import { rolldownAstroFrontmatterScanPlugin } from "./rolldown-plugin-astro-frontmatter.js";
import { getParts } from "./utils/generate-routes-json.js";
import { buildAssetsHeadersContent } from "./utils/headers.js";
import {
  hasUserImageService,
  normalizeImageServiceConfig,
  setImageConfig
} from "./utils/image-config.js";
import { createConfigPlugin } from "./vite-plugin-config.js";
import { createNodePrerenderPlugin } from "./vite-plugin-dev-server-prerender-middleware.js";
import {
  cloudflareConfigCustomizer,
  DEFAULT_SESSION_KV_BINDING_NAME,
  DEFAULT_IMAGES_BINDING_NAME
} from "./wrangler.js";
import { sessionDrivers } from "astro/config";
import { createCloudflarePrerenderer } from "./prerenderer.js";
import cfPrismPlugin from "./vite-plugin-prism.js";
import { loadWranglerEnv } from "./utils/wrangler-config.js";
const CLOUDFLARE_KV_SESSION_DRIVER_ENTRYPOINT = sessionDrivers.cloudflareKVBinding().entrypoint;
function usesCloudflareKVSessionDriver(session) {
  const driver = session?.driver;
  if (!driver) {
    return false;
  }
  if (typeof driver === "string") {
    return driver === "cloudflareKVBinding" || driver === "cloudflare-kv-binding";
  }
  const entrypoint = typeof driver.entrypoint === "string" ? driver.entrypoint : driver.entrypoint.toString();
  return entrypoint === CLOUDFLARE_KV_SESSION_DRIVER_ENTRYPOINT || entrypoint.endsWith("cloudflare-kv-binding");
}
function hasContentCollectionsConfig(srcDir) {
  const contentConfigPaths = [
    "content.config.mjs",
    "content.config.js",
    "content.config.mts",
    "content.config.ts",
    "content/config.mjs",
    "content/config.js",
    "content/config.mts",
    "content/config.ts",
    "live.config.mjs",
    "live.config.js",
    "live.config.mts",
    "live.config.ts"
  ];
  return contentConfigPaths.some((configPath) => existsSync(new URL(`./${configPath}`, srcDir)));
}
function resolveImageServiceEntrypoint(entrypoint, root) {
  if (entrypoint.startsWith(".")) {
    return new URL(entrypoint, root).href;
  }
  return entrypoint;
}
function createIntegration({
  imageService,
  sessionKVBindingName = DEFAULT_SESSION_KV_BINDING_NAME,
  imagesBindingName = DEFAULT_IMAGES_BINDING_NAME,
  prerenderEnvironment = "workerd",
  ...cloudflareOptions
} = {}) {
  let _config;
  let _buildOutput;
  let _originalClientDir;
  let _routes;
  let cfPluginConfig;
  let hasUserBuildImageService = false;
  let compileImageConfig = null;
  const { buildService, runtimeService } = normalizeImageServiceConfig(imageService);
  const needsImagesBinding = runtimeService === "cloudflare-binding";
  const hasBuildImageService = buildService === "compile" || buildService === "custom";
  return {
    name: "@astrojs/cloudflare",
    hooks: {
      "astro:config:setup": async ({ command, config, updateConfig, logger, addWatchFile }) => {
        if (!!process.versions.webcontainer) {
          throw new Error("`workerd` does not run on Stackblitz.");
        }
        let session = config.session;
        const isCompile = buildService === "compile";
        if (needsImagesBinding) {
          logger.info(
            `Enabling image processing with Cloudflare Images for production with the "${imagesBindingName}" Images binding.`
          );
        } else if (hasBuildImageService) {
          logger.info(
            `Enabling compile-time image optimization. Images will be pre-optimized at build time.`
          );
        }
        if (!session?.driver) {
          logger.info(
            `Enabling sessions with Cloudflare KV with the "${sessionKVBindingName}" KV binding.`
          );
          session = {
            driver: sessionDrivers.cloudflareKVBinding({
              binding: sessionKVBindingName
            }),
            cookie: session?.cookie,
            ttl: session?.ttl
          };
        }
        const needsSessionKVBinding = usesCloudflareKVSessionDriver(session);
        const needsImagesBindingForDev = isCompile && command === "dev";
        const usesContentCollections = hasContentCollectionsConfig(config.srcDir);
        const prebundleContentRuntime = command === "dev" && usesContentCollections;
        const isTypeGenPhase = command === "build" || command === "sync";
        const needsWorkerCache = config.cache?.provider?.name === "cloudflare";
        const adapterPluginConfig = {
          config: cloudflareConfigCustomizer({
            needsSessionKVBinding,
            sessionKVBindingName,
            imagesBindingName: needsImagesBinding || needsImagesBindingForDev ? imagesBindingName : false,
            needsWorkerCache
          }),
          ...prerenderEnvironment === "workerd" && {
            experimental: {
              prerenderWorker: {
                config(_, { entryWorkerConfig }) {
                  const { queues, ...restWorkerConfig } = entryWorkerConfig;
                  return {
                    ...restWorkerConfig,
                    name: "prerender",
                    ...queues?.producers?.length && {
                      queues: { producers: queues.producers }
                    },
                    ...needsImagesBinding && !restWorkerConfig.images && {
                      images: { binding: imagesBindingName }
                    }
                  };
                }
              }
            }
          }
        };
        cfPluginConfig = { ...cloudflareOptions, ...adapterPluginConfig };
        if (command === "preview") {
          globalThis.astroCloudflareConfig = cfPluginConfig;
        }
        const prismFiles = [
          "@astrojs/prism > prismjs",
          "@astrojs/prism > prismjs/components.js",
          "@astrojs/prism > prismjs/dependencies.js"
        ];
        const isAstroPrismPackageInstalled = await getIsAstroPrismInstalled(config.root);
        const userOptimizeDeps = config.vite?.optimizeDeps;
        const cloudflareVitePlugins = cfVitePlugin({
          ...cfPluginConfig,
          viteEnvironment: { name: "ssr" },
          assetsOnly: () => _buildOutput === "static"
        });
        if (isTypeGenPhase) {
          for (const plugin of cloudflareVitePlugins) {
            plugin.configureServer = void 0;
          }
        }
        updateConfig({
          build: {
            redirects: false
          },
          session,
          vite: {
            plugins: [
              ...prerenderEnvironment === "node" && command === "dev" ? [createNodePrerenderPlugin()] : [],
              cloudflareVitePlugins,
              {
                name: "@astrojs/cloudflare:cf-imports",
                enforce: "pre",
                resolveId: {
                  filter: {
                    id: /^cloudflare:/
                  },
                  handler(id) {
                    return { id, external: true };
                  }
                }
              },
              {
                name: "@astrojs/cloudflare:environment",
                configEnvironment(environmentName, _options) {
                  if (isTypeGenPhase) {
                    return { optimizeDeps: { noDiscovery: true, include: [] } };
                  }
                  const isServerEnvironment = ["astro", "ssr", "prerender"].includes(
                    environmentName
                  );
                  if (isServerEnvironment && !_options.optimizeDeps?.noDiscovery) {
                    return {
                      optimizeDeps: {
                        include: [
                          "@astrojs/cloudflare/image-service-workerd",
                          "@astrojs/cloudflare/entrypoints/server",
                          "astro",
                          "astro/runtime/**",
                          "astro > html-escaper",
                          "astro > mrmime",
                          "astro > zod/v4",
                          "astro > zod/v4/core",
                          "astro > clsx",
                          "astro > cookie",
                          "astro > devalue",
                          "astro > @oslojs/encoding",
                          "astro > es-module-lexer",
                          "astro > unstorage",
                          "astro > neotraverse/modern",
                          "astro > piccolore",
                          "astro > picomatch",
                          "astro/app",
                          "astro/app/fetch/default-handler",
                          "astro/fetch",
                          "astro/hono",
                          "astro/assets",
                          "astro/assets/runtime",
                          "astro/assets/utils/inferRemoteSize.js",
                          "astro/assets/fonts/runtime.js",
                          ...prebundleContentRuntime ? ["astro/content/runtime"] : [],
                          "astro/compiler-runtime",
                          "astro/jsx-runtime",
                          "astro/app/entrypoint/dev",
                          "astro/virtual-modules/middleware.js",
                          "astro/virtual-modules/transitions.js",
                          "astro/virtual-modules/transitions-events.js",
                          "astro/virtual-modules/transitions-router.js",
                          "astro/virtual-modules/transitions-swap-functions.js",
                          "astro/virtual-modules/transitions-types.js",
                          "astro/components",
                          ...isAstroPrismPackageInstalled ? prismFiles : [],
                          ...Array.isArray(userOptimizeDeps?.include) ? userOptimizeDeps.include : []
                        ],
                        exclude: [
                          "unstorage/drivers/cloudflare-kv-binding",
                          "astro:*",
                          "virtual:astro:*",
                          "virtual:astro-cloudflare:*",
                          "virtual:@astrojs/*",
                          "@astrojs/starlight",
                          ...Array.isArray(userOptimizeDeps?.exclude) ? userOptimizeDeps.exclude : []
                        ],
                        rolldownOptions: {
                          plugins: [rolldownAstroFrontmatterScanPlugin()]
                        }
                      }
                    };
                  } else if (environmentName === "client") {
                    return {
                      optimizeDeps: {
                        include: ["astro/runtime/client/dev-toolbar/entrypoint.js"],
                        // Workaround for https://github.com/vitejs/vite/issues/20867
                        // When dependencies are discovered mid-request (e.g. a linked package
                        // used with client:only), concurrent requests can fail with 504 because
                        // the dep optimizer's metadata object gets replaced during `await info.processing`.
                        ignoreOutdatedRequests: true
                      }
                    };
                  }
                }
              },
              {
                enforce: "post",
                name: "@astrojs/cloudflare:cf-externals",
                applyToEnvironment: (environment) => environment.name === "ssr" || environment.name === "prerender",
                config(conf) {
                  if (conf.ssr) {
                    conf.ssr.external = void 0;
                  }
                }
              },
              createConfigPlugin({
                sessionKVBindingName,
                // `imageServiceEntrypoint` is finalized in `astro:config:done`:
                // integrations may set `image.service` via `updateConfig()` after
                // this hook runs (the adapter always runs first), so the service
                // cannot be resolved yet. The plugin serializes this object lazily
                // at load time, after the mutation below has happened.
                compileImageConfig: hasBuildImageService && command !== "dev" ? compileImageConfig = {
                  base: config.base,
                  assetsPrefix: typeof config.build.assetsPrefix === "string" ? config.build.assetsPrefix : void 0,
                  imageServiceEntrypoint: "@astrojs/cloudflare/image-service-workerd",
                  buildAssets: config.build.assets ?? "_astro"
                } : null,
                cacheProviderEnabled: needsWorkerCache
              }),
              cfPrismPlugin()
            ]
          },
          image: setImageConfig(imageService, config.image, command, logger)
        });
        if (cloudflareOptions.configPath) {
          addWatchFile(new URL(cloudflareOptions.configPath, config.root));
        }
        addWatchFile(new URL("./wrangler.toml", config.root));
        addWatchFile(new URL("./wrangler.json", config.root));
        addWatchFile(new URL("./wrangler.jsonc", config.root));
      },
      "astro:routes:resolved": ({ routes }) => {
        _routes = routes;
      },
      "astro:config:done": ({ setAdapter, config, injectTypes, logger, buildOutput }) => {
        _config = config;
        _buildOutput = buildOutput;
        _originalClientDir = new URL(config.build.client.href);
        hasUserBuildImageService = hasBuildImageService && hasUserImageService(config.image);
        if (compileImageConfig && hasUserBuildImageService) {
          compileImageConfig.imageServiceEntrypoint = config.image.service.entrypoint;
        }
        if (config.base !== "/") {
          config.build.client = new URL("." + config.base + "/", config.build.client);
        }
        injectTypes({
          filename: "cloudflare.d.ts",
          content: '/// <reference types="@astrojs/cloudflare/types.d.ts" />'
        });
        setAdapter({
          name: "@astrojs/cloudflare",
          adapterFeatures: {
            buildOutput,
            middlewareMode: "classic",
            preserveBuildClientDir: true,
            preserveBuildServerDir: true
          },
          entrypointResolution: "auto",
          previewEntrypoint: "@astrojs/cloudflare/entrypoints/preview",
          supportedAstroFeatures: {
            serverOutput: "stable",
            hybridOutput: "stable",
            staticOutput: "stable",
            i18nDomains: "experimental",
            sharpImageService: {
              support: "limited",
              message: "When using a custom image service, ensure it is compatible with the Cloudflare Workers runtime.",
              // Only 'custom' could potentially use sharp at runtime.
              suppress: buildService === "custom" ? "default" : "all"
            },
            envGetSecret: "stable"
          }
        });
        loadWranglerEnv(config.root, cloudflareOptions.configPath, logger);
      },
      "astro:build:start": ({ setPrerenderer }) => {
        if (prerenderEnvironment === "workerd") {
          setPrerenderer(
            createCloudflarePrerenderer({
              root: _config.root,
              serverDir: _config.build.server,
              clientDir: _config.build.client,
              base: _config.base,
              trailingSlash: _config.trailingSlash,
              cfPluginConfig,
              hasBuildImageService,
              userImageServiceEntrypoint: hasUserBuildImageService ? resolveImageServiceEntrypoint(_config.image.service.entrypoint, _config.root) : void 0
            })
          );
        }
      },
      "astro:build:setup": ({ vite, target }) => {
        if (target === "server") {
          vite.resolve ||= {};
          vite.resolve.alias ||= {};
          vite.ssr ||= {};
          vite.ssr.noExternal = true;
          vite.build ||= {};
          vite.build.rolldownOptions ||= {};
          vite.build.rolldownOptions.output ||= {};
          vite.build.rolldownOptions.external = ["sharp"];
          vite.build.rolldownOptions.output.banner ||= "globalThis.process ??= {}; globalThis.process.env ??= {};";
          vite.define = {
            "process.env": "process.env",
            "globalThis.__ASTRO_IMAGES_BINDING_NAME": JSON.stringify(imagesBindingName),
            ...vite.define
          };
        }
      },
      "astro:build:done": async ({ dir, logger, assets }) => {
        if (_config.base !== "/") {
          for (const file of [".assetsignore", "_headers", "_redirects"]) {
            try {
              await rename(
                new URL(`./${file}`, _config.build.client),
                new URL(`./${file}`, _originalClientDir)
              );
            } catch {
            }
          }
          try {
            const wranglerJsonUrl = new URL("./wrangler.json", _config.build.server);
            const raw = await readFile(wranglerJsonUrl, "utf-8");
            const wranglerConfig = JSON.parse(raw);
            if (wranglerConfig.assets?.directory) {
              wranglerConfig.assets.directory = normalizePath(
                relative(fileURLToPath(_config.build.server), fileURLToPath(_originalClientDir))
              );
              await writeFile(wranglerJsonUrl, JSON.stringify(wranglerConfig));
            }
          } catch {
          }
        }
        if (_config.build.assetsPrefix) {
          logger.debug(
            "Skipping Cache-Control injection for assets \u2014 `build.assetsPrefix` is set, so assets are served from a different origin."
          );
        } else {
          const headersPath = new URL("./_headers", _originalClientDir);
          const result = await buildAssetsHeadersContent(
            {
              assetsDir: _config.build.assets,
              basePrefix: removeTrailingForwardSlash(_config.base),
              headersPath
            },
            (path) => readFile(path, "utf-8")
          );
          if (result === null) {
            logger.debug(
              `Skipping Cache-Control injection \u2014 _headers already sets Cache-Control on a matching rule.`
            );
          } else {
            const tempPath = new URL("./_headers.tmp", _originalClientDir);
            try {
              await writeFile(tempPath, result.content);
              await rename(tempPath, headersPath);
            } catch (err) {
              await unlink(tempPath).catch(() => {
              });
              throw err;
            }
            logger.info(
              `Injected immutable Cache-Control for ${result.assetsPattern} into _headers.`
            );
          }
        }
        let redirectsExists = false;
        try {
          const redirectsStat = await stat(new URL("./_redirects", _originalClientDir));
          if (redirectsStat.isFile()) {
            redirectsExists = true;
          }
        } catch (_error) {
          redirectsExists = false;
        }
        const redirects = [];
        if (redirectsExists) {
          const rl = createInterface({
            input: createReadStream(new URL("./_redirects", _originalClientDir)),
            crlfDelay: Number.POSITIVE_INFINITY
          });
          for await (const line of rl) {
            const parts = line.split(" ");
            if (parts.length >= 2) {
              const p = removeLeadingForwardSlash(parts[0]).split("/").filter(Boolean).map((s) => {
                const syntax = s.replace(/\/:.*?(?=\/|$)/g, "/*").replace(/\?.*$/, "");
                return getParts(syntax);
              });
              redirects.push(p);
            }
          }
        }
        const trueRedirects = createRedirectsFromAstroRoutes({
          config: _config,
          routeToDynamicTargetMap: new Map(
            Array.from(
              _routes.filter((route) => route.type === "redirect").map((route) => [route, ""])
            )
          ),
          dir,
          buildOutput: _buildOutput,
          assets
        });
        if (!trueRedirects.empty()) {
          try {
            await appendFile(
              new URL("./_redirects", _originalClientDir),
              printAsRedirects(trueRedirects)
            );
          } catch (_error) {
            logger.error("Failed to write _redirects file");
          }
        }
        delete process.env.CLOUDFLARE_VITE_BUILD;
      }
    }
  };
}
async function getIsAstroPrismInstalled(rootURL) {
  try {
    const pkgURL = new URL("./package.json", rootURL);
    const input = await readFile(pkgURL, { encoding: "utf-8" });
    const pkgJson = JSON.parse(input);
    return Object.hasOwn(pkgJson["dependencies"], "@astrojs/prism");
  } catch {
    return false;
  }
}
export {
  createIntegration as default
};
